Functionaity ready

tcorea opened the pull request:
I have created the Poker Evaluator class that evaluates a the users hand and determines its value whether it is a royal flush or pair, etc. It will be used by the PokerChart class which has most/common poker odds that exist in the game. The poker calculator gets information from the user about its hand and community board and tries to then calculates the probability using outs, loss, and  unkowns and turning it into a ratio and then converts that into a percentage which is our output. Tostring method will have output. but must call calculations method to get it.
I was able to come with this based off the links given to me about the poker chart. Also, I want try to make it more intense with a complex algorithm hopefully for final. two junit test cases available. 
~Thomas Corea

tcorea committed:
Functionaity ready

I have created the Poker Evaluator class that evaluates a the users hand and determines its value whether it is a royal flush or pair, etc. It will be used by the PokerChart class which has most/common poker odds that exist in the game. The poker calculator gets information from the user about its hand and community board and tries to then calculates the probability using outs, loss, and  unkowns and turning it into a ratio and then converts that into a percentage which is our output. Tostring method will have output. but must call calculations method to get it.

vlyandra commented on src/main/java/calculator/PokerChart.java:
```diff
@@ -0,0 +1,374 @@
+import java.util.Collections;
+import java.util.ListIterator;
+
+/**
```
Javadocs must be the beginning of the file.

tcorea replied:
I thought you said earlier that packages go at the top first ? sorry

vlyandra replied:
Package declarations must be the first source line. Javadocs go above.

vlyandra replied:
Package declarations must be the first source line. Javadocs go above.

vlyandra commented on src/main/java/calculator/PokerChart.java:
```diff
@@ -0,0 +1,374 @@
+  /**
+   * It will check the inside straight draw probability.
+   * 
+   * @return boolean value
```
@return statements should be descriptive of what they're returning. `value` doesn't tell the next person looking at this code what the return means.

tcorea replied:
I made @return more descriptive

vlyandra replied:
Where?

vlyandra replied:
Where?

vlyandra commented on src/main/java/calculator/PokerChart.java:
```diff
@@ -0,0 +1,374 @@
+   * @return
+   */
+  public boolean hasNoPairToPair(ArrayList<Cards> userCards, ArrayList<Cards> communityCards) {
+    if (!(evaluator.hasPair(userCards, communityCards)))
```
you can just `return !(evaluator.hasPair(userCards, communityCards))` here

tcorea replied:
I repaired the unnecessary coding.

vlyandra commented on src/main/java/calculator/PokerChart.java:
```diff
@@ -0,0 +1,374 @@
+   * @return boolean value.
+   */
+  public boolean hasTwoPair2House(ArrayList<Cards> userCards, ArrayList<Cards> communityCards) {
+    if (evaluator.hasTwoPairs(userCards, communityCards)) {
```
This can be replaced with `return evaluator.hasTwoPairs(userCards, communityCards)`

tcorea replied:
I made the adjustments to the statements above.

vlyandra commented on src/main/java/calculator/PokerChart.java:
```diff
@@ -0,0 +1,374 @@
+    // now we have the high card which is in the hand lets check if that hand is highest in the
+    // community
+    // check both cards
+    for (Cards card : communityCards) {
```
These two for loops could be combined, since you're returning if either of them is true at any point.

vlyandra commented on src/main/java/calculator/PokerEvaluator.java:
```diff
@@ -0,0 +1,447 @@
+   * @return
+   */
+  public boolean hasPair(ArrayList<Cards> userCards, ArrayList<Cards> communityCards) {
+    int num_OfPairs, num_OfPairs1;
```
Java variable names should be lowerCamelCase, no underscores.

vlyandra commented on src/main/java/calculator/PokerEvaluator.java:
```diff
@@ -0,0 +1,447 @@
+  public boolean hasStraightFlush(ArrayList<Cards> userCards, ArrayList<Cards> communityCards) {
+
+    // does the user have a straight.
+    if (hasStraight(userCards, communityCards) && hasFlush(userCards, communityCards))
```
`return hasStraight(userCards, communityCards) && hasFlush(userCards, communityCards)` is sufficient here

vlyandra commented on src/main/java/calculator/PokerEvaluator.java:
```diff
@@ -0,0 +1,447 @@
+   * @return boolean value
+   */
+  public boolean hasfullHouse(ArrayList<Cards> userCards, ArrayList<Cards> communityCards) {
+    if (hasThreeOfKind(userCards, communityCards) && hasPair(userCards, communityCards))
```
See above, `return hasThreeOfKind(userCards, communityCards) && hasPair(userCards, communityCards)`

vlyandra commented on src/main/java/calculator/PokerEvaluator.java:
```diff
@@ -0,0 +1,447 @@
+  }
+
+  /**
+   * Checks to see if you have a pair.
```
This javadoc appears to be incorrect. This checks for two pairs, yes?

vlyandra commented on src/main/java/calculator/PokerChart.java:
```diff
@@ -0,0 +1,374 @@
+   * @return boolean value
+   */
+  public boolean hasPocketPair(ArrayList<Cards> userCards, ArrayList<Cards> communityCards) {
+    if (evaluator.hasPair(userCards, communityCards))
```
This doesn't assure a pocket pair, since `evaluator.hasPair` will return true if the cards on the board also allow for a pair.

vlyandra commented on src/main/java/calculator/PokerEvaluator.java:
```diff
@@ -0,0 +1,447 @@
+    int sameKind2 = 0;
+    // Comparing hand for pairs/match
+    if (userCards.get(0).getRank().getRankValue() == userCards.get(1).getRank().getRankValue())
+      return false;// impossible to have two pairs if your only two cards are same
```
But you can have two pairs if there's a pair on the board and you have a pocket pair, right?

vlyandra commented on src/main/java/calculator/PokerEvaluator.java:
```diff
@@ -0,0 +1,447 @@
+          suitCounter++;
+
+      }
+      if (suitCounter == 5)
```
While unlikely it is possible that there can be more than 5 cards of the same suit within the 7 available cards.

vlyandra commented on src/main/java/calculator/PokerEvaluator.java:
```diff
@@ -0,0 +1,447 @@
+   */
+  public boolean hasRoyalFlush(ArrayList<Cards> userCards, ArrayList<Cards> communityCards) {
+    // boolean representation of the cards
+    boolean aceCard, kingCard, queenCard, jackCard, tenCard;// royal flush
```
A data structure would be nice here instead of 5 separate variables.

vlyandra commented on src/test/java/calculator/PokerCalculatorTest.java:
```diff
@@ -0,0 +1,412 @@
+
+  private Cards card1, card2, card3, card4, card5, card6, card7;
+
+  @BeforeClass
```
We use JUnit 5, please use JUnit 5 terminology.

vlyandra commented on src/test/java/calculator/PokerCalculatorTest.java:
```diff
@@ -0,0 +1,412 @@
+  }
+
+  @Before
+  public void setUp() throws Exception {
```
It looks like you're using mostly the same cards everywhere, wouldn't it be easier to just make them here? You're also doing a reset before every test, that could happen here too.

vlyandra commented on src/test/java/calculator/PokerEvaluatorTest.java:
```diff
@@ -0,0 +1,567 @@
+  }
+
+  @Test
+  public void testHasPair() {
```
Is there a reason you don't reset here?

vlyandra commented on src/test/java/calculator/PokerEvaluatorTest.java:
```diff
@@ -0,0 +1,567 @@
+    assertTrue(calcEval.hasTwoPairs(calcObj.getUserCards(), calcObj.getCommunityCards()));
+
+
+    // another test
```
If it's another test, it should be written as another, separate test.

vlyandra commented on src/test/java/calculator/PokerCalculatorTest.java:
```diff
@@ -0,0 +1,412 @@
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
```
There seem to be no assertions used at all. Is there a reason these tests don't test anything?

vlyandra commented on src/test/java/calculator/PokerEvaluatorTest.java:
```diff
@@ -0,0 +1,567 @@
+
+  @Before
+  public void setUp() throws Exception {}
+
```
Again, the card resets could all be done here or in @AfterEach

vlyandra commented on src/test/java/calculator/PokerEvaluatorTest.java:
```diff
@@ -0,0 +1,567 @@
+    assertFalse(calcEval.hasHighCard(calcObj.getUserCards(), calcObj.getCommunityCards()));// no
+                                                                                           // highcard
+
+    calcObj.resetUserHand();
```
This is two tests, right? They should be separated out into separate tests.

tcorea replied:
This will be resolved later since i have to find a new IDE

vlyandra commented on src/test/java/calculator/PokerEvaluatorTest.java:
```diff
@@ -0,0 +1,567 @@
+  private Cards userCard1;
+  private Cards userCard2;
+  private PokerEvaluator calcEval = new PokerEvaluator();
+  private PokerCalculator calcObj = new PokerCalculator();
```
If looks to me like you're only using calcObj to hold cards here? It's preferable to only use outside classes that are strictly necessary, and this creates an unnecessary dependency.

tcorea replied:
will change later once i update to Junit 5

vlyandra commented on src/test/java/calculator/PokerCalculatorTest.java:
```diff
@@ -0,0 +1,412 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
```
We use JUnit 5, please use JUnit 5 packages from `org.junit.jupiter.api`

tcorea replied:
Ok I will have to re-install entire different IDE. So ill work on that.

vlyandra replied:
Your IDE should have nothing to do with this. Maven automatically manages the JUnit 5 dependency.

tcorea replied:
> Your IDE should have nothing to do with this. Maven automatically manages the JUnit 5 dependency.

Well im using Eclipse NEon so im not sure what you mean by this statement. According to the Eclipse website I have to install Eclipse Oxygen or later version

vlyandra commented on src/test/java/calculator/PokerEvaluatorTest.java:
```diff
@@ -0,0 +1,567 @@
+    calcObj.addCommunityCard(card4);
+    calcObj.addCommunityCard(card5);
+
+    assertTrue(calcEval.hasPair(calcObj.getUserCards(), calcObj.getCommunityCards()));
```
If you are making multiple test cases in a test, they need to be separate tests.

tcorea replied:
will resolve when i update eclipse

vlyandra commented on src/main/java/calculator/PokerCalculator.java:
```diff
@@ -12,580 +12,246 @@
-}
+ public class PokerCalculator {
+   // THe value of the hand.
+   private int handValue;// 1-10 . 10 being the most valuable
```
Where is this used, other than assigned in the constructor?

tcorea replied:
it has been removed.

vlyandra reviewed (changes requested):
- Please lint your code, especially regarding bracket use/placement.
- A lot of your method Javadocs do not have any @param information, and the @return information is very vague
- I'm not sure what's going on with PokerCalculatorTest. Even if you're going for TDD they should still test something, even if the tests don't all pass yet (at which point, they shouldn't be included as part of the test suite)

vlyandra commented:
Also, this branch is way behind master. Please resolve the conflicts.

tcorea committed:
Changes completed

I made changes you requested except for the Junit test cases. I will need time to re do them in a different IDE because the one im using doesn't support JUNIT 5 . I changed the Javadocs and corrected it with more details about parameters and returns. I modified the code by getting rid of unncessary coding.

vlyandra reviewed (changes requested):
Please at least address all of the requested changes. There are some still unresolved.

Also, your branch conflicts hard with master. Please resolve the merge conflicts.

tcorea commented:
> Please at least address all of the requested changes. There are some still unresolved.
> 
> Also, your branch conflicts hard with master. Please resolve the merge conflicts.

ok I think everything is resolved for now.

vlyandra commented:
There are unresolved change requests still, eg https://github.com/vlyandra/PokerProjectCapstone/pull/19#discussion_r280775201 and the branch still had serious conflicts with master

tcorea commented:
> There are unresolved change requests still, eg [#19 (comment)](https://github.com/vlyandra/PokerProjectCapstone/pull/19#discussion_r280775201) and the branch still had serious conflicts with master

What does that mean ? I just made an adjustment. So I think its resolved. I found some instructions online. This is my first time using github for project or whatever so bare with me. THanks

vlyandra commented:
Ah, I hadn’t caught your most recent commit.

tcorea commented:
> Ah, I hadn’t caught your most recent commit.

no worries did that fix it ?

vlyandra commented:
It’ll merge, but somehow you’ve touched every file and you’re actually reverting some previous progress. I can’t approve this without losing functionality.

vlyandra commented:
Did you check the link in the README with git instructions?

tcorea commented:
> Did you check the link in the README with git instructions?
yes I did it read it. So do I cancel this merge and just do another one ?

tcorea commented:
> It’ll merge, but somehow you’ve touched every file and you’re actually reverting some previous progress. I can’t approve this without losing functionality.

should i close this pull request?

tcorea committed:
Updated to Master to resolve issue

Issue should be resolved

tcorea committed:
JUnit

Adding Junit test casea

tcorea committed:
Logical Issues Resolved

I resolved the logical issues requested.  Changed Pocket Pair conditions. Reevaluated the hasPair method. and many others.

vlyandra commented on src/main/java/calculator/PokerEvaluator.java:
```diff
@@ -11,463 +11,483 @@
+           suitCounter++;
+
+       }
+       if (suitCounter == 5)
```
A user can have a flush with their hand and all of the community cards by having 5 *or more* of the same suit total.

vlyandra committed:
Clean up PokerEvaluator and correct invalid logic

vlyandra reviewed (approved):
Fixed most of the remaining issues I found. Approving so we can get this merged.
